client/core/core.go

```diff
@@ -6116,80 +6145,76 @@ func (c *Core) dbTrackers(dc *dexConnection) (map[order.OrderID]*trackedTrade, e
-// dexConnection. If there are active trades, the necessary wallets will be
-// unlocked. To prevent spamming wallet connections, the 'failed' map will be
-// populated with asset IDs for which the attempt to connect or unlock has
-// failed. The failed map should be passed on subsequent calls for other dexes.
-func (c *Core) loadDBTrades(dc *dexConnection, crypter encrypt.Crypter, failed map[uint32]struct{}) ([]*trackedTrade, error) {
-	// Parse the active trades and see if any wallets need unlocking.
-	trades, err := c.dbTrackers(dc)
+// loadDBTrades load's the active trades from the db, populates the trade's
+// wallets field and some other metadata, and adds the trade to the
+// dexConnection's trades map. readyToTick will still be set to false, so
+// resumeTrades must be run before the trades will be processed.
+func (c *Core) loadDBTrades(dc *dexConnection) error {
+	trackers, err := c.dbTrackers(dc)
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving active matches: %w", err)
+		return fmt.Errorf("error retrieving active matches: %w", err)
 	}
 
 	errs := newErrorSet(dc.acct.host + ": ")
-	ready := make([]*trackedTrade, 0, len(trades))
-	for _, trade := range trades {
-		if !trade.isActive() {
+	var tradesLoaded uint32
+	for _, tracker := range trackers {
+		if !tracker.isActive() {
 			// In this event, there is a discrepancy between the active criteria
 			// between dbTrackers and isActive that should be resolved.
-			c.log.Warnf("Loaded inactive trade %v from the DB.", trade.ID())
+			c.log.Warnf("Loaded inactive trade %v from the DB.", tracker.ID())
 			continue
 		}
-		base, quote := trade.Base(), trade.Quote()
-		_, baseFailed := failed[base]
-		_, quoteFailed := failed[quote]
-		if !baseFailed {
-			baseWallet, err := c.connectedWallet(base)
-			if err != nil {
-				baseFailed = true
-				failed[base] = struct{}{}
-				c.log.Errorf("Connecting to wallet %s failed: %v", unbip(base), err)
-			} else if !baseWallet.unlocked() {
-				err = baseWallet.Unlock(crypter)
-				if err != nil {
-					baseFailed = true
-					failed[base] = struct{}{}
-					c.log.Errorf("Unlock wallet %s failed: %v", unbip(base), err)
-				}
-			}
+
+		trade := tracker.Trade()
+
+		walletSet, assetConfigs, versCompat, err := c.walletSet(dc, tracker.Base(), tracker.Quote(), trade.Sell)
+		if err != nil {
+			errs.add("failed to load wallets for trade ID %s: %w", tracker.ID(), err)
+			continue
 		}
-		if !baseFailed && !quoteFailed {
-			quoteWallet, err := c.connectedWallet(quote)
-			if err != nil {
-				quoteFailed = true
-				failed[quote] = struct{}{}
-				c.log.Errorf("Connecting to wallet %s failed: %v", unbip(quote), err)
-			} else if !quoteWallet.unlocked() {
-				err = quoteWallet.Unlock(crypter)
-				if err != nil {
-					quoteFailed = true
-					failed[quote] = struct{}{}
-					c.log.Errorf("Unlock wallet %s failed: %v", unbip(quote), err)
-				}
+
+		tracker.wallets = walletSet
+		dc.trades[tracker.ID()] = tracker
```

Comment:
I haven't digested the logic yet as it pertains to the `resumeTrades bool`, but we should be sure never to overwrite an existing map entry.  If there are two `trackedTrade` instances for the same orders, bad things can and will happen.

Reply:
I don't think an overwrite could happen because this function is only called at login, but I've added a check to make sure.

The `resumeTrades bool` arguments have anything to do with adding trades to the map. It's only to avoid recursive calls back to `resumeTrades` when wallets are being unlocked in `resumeTrade`.

Reply:
In `(*Core).resumeTrade`, there is a large amount of work being done even for trades that don't need to be resumed i.e. are already running fine.  We should really leave alone trades that are fine.  Calling FindingCoins repeatedly, messing with account-based wallet reserves, etc.  Need to avoid all that somehow.  It looks to me like all orders that would load fine are going to go through all this, return true from `resumeTrade`, and emit an order note.

We talked about having the lame trades in the trades map for reasons, but that doesn't mean we can't add a tracking structure as a field to `dexConnections` to assist.

That's potentially separate, but I also thing we should look at the need for the bool because it does hint that something is amiss in the design.

A related thing that stands out is when wallet A is unlocked, we try to resume all trades even if they have nothing to do with wallet A.

Reply:
Ohh, `readyToTick` blocks calling these again?  OK.  Please consider the other points though.

Reply:
> Seems like resumeTrade could check this too and return true.

Good call, I'll add a check. 

> That's potentially separate, but I also think we should reconsider at the need for the bool because it does hint that something is amiss in the design.

I was considering putting the lame trades into a separate map, and then having `cancelOrder` search through those as well, but it seemed cleaner to do it with the bool. Why is this a bad design? The `trades` map has all of the active trades for a dex, but some of them could not yet complete pre-processing, so we mark this with a flag. It seems decently straightforward.

> A related thing that stands out is when wallet A is unlocked, we try to resume all trades even if they have nothing to do with wallet A.

I considered this also.. we could pass an assetID filter to `resumeTrades`, but the max cost of not doing it is one failed call to per wallet that has active trades.
Maybe it's far fetched, but there could be a case where the user has an active trade with asset A-B, and one with A-C, and both A and B initially failed to connect. Then they start both the A and B RPC wallets, unlock the B wallet, and A is able to automatically unlock. If only resume trades involving wallet B, then the A-C trade will not be resumed. Unless we allow recursive calls back to `resumeTrades`.